k2/python/tests/autograd_test.py

```diff
@@ -0,0 +1,294 @@
+
+        fsa1.scores.requires_grad_(True)
+        fsa2.scores.requires_grad_(True)
+        fsa3.scores.requires_grad_(True)
+
+        fsa_vec = k2.create_fsa_vec([fsa1, fsa2, fsa3])
+
+        log_like = k2.get_tot_scores(fsa_vec,
+                                     log_semiring=False,
+                                     use_float_scores=True)
+        assert log_like.dtype == torch.float32
+        expected_log_like = torch.tensor([14, 13, 105.5])
+        assert torch.allclose(log_like, expected_log_like)
+
+        log_like.sum().backward()
+
+        fsa1_best_arc_indexes = torch.tensor([1, 3, 5, 10])
+        assert torch.allclose(fsa1.scores.grad[fsa1_best_arc_indexes],
+                              torch.ones(4, dtype=torch.float32))
+        assert fsa1.scores.grad.sum() == 4
+
+        fsa2_best_arc_indexes = torch.tensor([1, 4, 5, 7])
+        assert torch.allclose(fsa2.scores.grad[fsa2_best_arc_indexes],
+                              torch.ones(4, dtype=torch.float32))
+        assert fsa2.scores.grad.sum() == 4
+
+        fsa3_best_arc_indexes = torch.tensor([1, 3])
+        assert torch.allclose(fsa3.scores.grad[fsa3_best_arc_indexes],
+                              torch.ones(2, dtype=torch.float32))
+        assert fsa3.scores.grad.sum() == 2
+
+        # now for double
+        fsa1.scores.grad = None
+        fsa2.scores.grad = None
+        fsa3.scores.grad = None
+        log_like = k2.get_tot_scores(fsa_vec,
+                                     log_semiring=False,
+                                     use_float_scores=False)
+
+        assert log_like.dtype == torch.float64
+        expected_log_like = expected_log_like.to(torch.float64)
+        assert torch.allclose(log_like, expected_log_like)
+
+        log_like.sum().backward()
+
+        assert torch.allclose(fsa1.scores.grad[fsa1_best_arc_indexes],
+                              torch.ones(4, dtype=torch.float32))
+        assert fsa1.scores.grad.sum() == 4
+
+        assert torch.allclose(fsa2.scores.grad[fsa2_best_arc_indexes],
+                              torch.ones(4, dtype=torch.float32))
+        assert fsa2.scores.grad.sum() == 4
+
+        assert torch.allclose(fsa3.scores.grad[fsa3_best_arc_indexes],
+                              torch.ones(2, dtype=torch.float32))
+        assert fsa3.scores.grad.sum() == 2
+
+    def test_log_single_fsa(self):
+        s = '''
+            0 1 1 0.1
+            0 2 2 0.2
+            1 2 3 0.3
+            1 3 4 0.4
+            2 3 5 0.5
+            3 4 -1 0
+            4
+        '''
+        fsa = k2.Fsa.from_str(s)
+        fsa.scores.requires_grad_(True)
+        fsa_vec = k2.create_fsa_vec([fsa])
+        log_like = k2.get_tot_scores(fsa_vec,
+                                     log_semiring=True,
+                                     use_float_scores=True)
+        assert log_like.dtype == torch.float32
+        # The expected_log_like is computed using gtn.
+        # See https://bit.ly/3oUiRx9
```

Comment:
what not just using the host version of `GetForwardScores` to compute this? (we have wrapped it to `csrc/host_shim.h`). BTW, for grad stuff, we may need to create a helper function with `GetTotScore()` (a host version should be fine) for (automatic) test purpose.

Reply:
If you are just talking about the  test code, I think we can just write extra tests later.

Reply:
I just wanted to compare that `k2` computes the same gradient values as `gtn` when
they are given the same input FSA.

Reply:
> If you are just talking about the test code, I think we can just write extra tests later.

Yeah, sure, what I mean is also not doing this now.

Reply:
> I just wanted to compare that `k2` computes the same gradient values as `gtn` when
> they are given the same input FSA.

What I thought is, if we use gtn for test purpose (not experiment purpose) in a low level, it may make people feel that the correctness of  k2 is relying on gtn. And I'm not even sure if gtn handles those `inifity` stuff correctly with fast-math.

But, anyway, it's OK for now to leave as-is, since this is just a simple test case.